File: apple/Sources/Autograph/Channel.swift
import Clibautograph
import Foundation

private func createHello() -> Bytes {
    createBytes(autograph_hello_size())
}

private func createIndex() -> Bytes {
    createBytes(autograph_index_size())
}

private func createSafetyNumber() -> Bytes {
    createBytes(autograph_safety_number_size())
}

private func createSecretKey() -> Bytes {
    createBytes(autograph_secret_key_size())
}

private func createSignature() -> Bytes {
    createBytes(autograph_signature_size())
}

private func createSize() -> Bytes {
    createBytes(autograph_size_size())
}

private func createStateBytes() -> Bytes {
    createBytes(autograph_state_size())
}

private func createCiphertext(_ plaintext: Bytes) -> Bytes {
    let size = autograph_ciphertext_size(plaintext.count)
    return createBytes(size)
}

private func createPlaintext(_ ciphertext: Bytes) -> Bytes {
    let size = autograph_plaintext_size(ciphertext.count)
    return createBytes(size)
}

private func createSessionCiphertext(_ state: Bytes) -> Bytes {
    let size = autograph_ciphertext_size(autograph_session_size(state))
    return createBytes(size)
}

private func readIndex(_ bytes: Bytes) -> UInt32 {
    autograph_read_index(bytes)
}

private func readSize(_ bytes: Bytes) -> Int {
    Int(autograph_read_size(bytes))
}

private func resizePlaintext(_ plaintext: Bytes, _ size: Bytes) -> Bytes {
    Array(plaintext[0 ..< readSize(size)])
}

public class State {
    public var bytes: Bytes

    init() {
        bytes = createStateBytes()
    }
}

public class Channel {
    var state: State

    init(state: State) {
        self.state = state
    }

    public func useKeyPairs(
        identityKeyPair: Bytes,
        ephemeralKeyPair: Bytes
    ) throws -> Bytes {
        var publicKeys = createHello()
        let success = autograph_use_key_pairs(
            &publicKeys,
            &state.bytes,
            identityKeyPair,
            ephemeralKeyPair
        )
        if !success {
            throw Error.initialization
        }
        return publicKeys
    }

    public func usePublicKeys(publicKeys: Bytes) {
        autograph_use_public_keys(&state.bytes, publicKeys)
    }

    public func authenticate() throws -> Bytes {
        var safetyNumber = createSafetyNumber()
        let success = autograph_authenticate(&safetyNumber, &state.bytes)
        if !success {
            throw Error.authentication
        }
        return safetyNumber
    }

    public func keyExchange(isInitiator: Bool) throws -> Bytes {
        var signature = createSignature()
        let success = autograph_key_exchange(
            &signature,
            &state.bytes,
            isInitiator
        )
        if !success {
            throw Error.keyExchange
        }
        return signature
    }

    public func verifyKeyExchange(signature: Bytes) throws {
        let success = autograph_verify_key_exchange(&state.bytes, signature)
        if !success {
            throw Error.keyExchange
        }
    }

    public func encrypt(plaintext: Bytes) throws -> (UInt32, Bytes) {
        var ciphertext = createCiphertext(plaintext)
        var index = createIndex()
        let success = autograph_encrypt_message(
            &ciphertext,
            &index,
            &state.bytes,
            plaintext,
            plaintext.count
        )
        if !success {
            throw Error.encryption
        }
        return (readIndex(index), ciphertext)
    }

    public func decrypt(message: Bytes) throws -> (UInt32, Bytes) {
        var plaintext = createPlaintext(message)
        var index = createIndex()
        var size = createSize()
        let success = autograph_decrypt_message(
            &plaintext,
            &size,
            &index,
            &state.bytes,
            message,
            message.count
        )
        if !success {
            throw Error.decryption
        }
        return (readIndex(index), resizePlaintext(plaintext, size))
    }

    public func certifyData(data: Bytes) throws -> Bytes {
        var signature = createSignature()
        let success = autograph_certify_data(
            &signature,
            &state.bytes,
            data,
            data.count
        )
        if !success {
            throw Error.certification
        }
        return signature
    }

    public func certifyIdentity() throws -> Bytes {
        var signature = createSignature()
        let success = autograph_certify_identity(
            &signature,
            &state.bytes
        )
        if !success {
            throw Error.certification
        }
        return signature
    }

    public func verifyData(
        data: Bytes,
        publicKey: Bytes,
        signature: Bytes
    ) -> Bool {
        autograph_verify_data(
            &state.bytes,
            data,
            data.count,
            publicKey,
            signature
        )
    }

    public func verifyIdentity(
        publicKey: Bytes,
        signature: Bytes
    ) -> Bool {
        autograph_verify_identity(
            &state.bytes,
            publicKey,
            signature
        )
    }

    public func close() throws -> (Bytes, Bytes) {
        var key = createSecretKey()
        var ciphertext = createSessionCiphertext(state.bytes)
        let success = autograph_close_session(&key, &ciphertext, &state.bytes)
        if !success {
            throw Error.session
        }
        return (key, ciphertext)
    }

    public func open(key: inout Bytes, ciphertext: Bytes) throws {
        let success = autograph_open_session(
            &state.bytes,
            &key,
            ciphertext,
            ciphertext.count
        )
        if !success {
            throw Error.session
        }
    }
}

File: android/src/main/java/sh/autograph/Channel.kt
package sh.autograph

class Channel(private val state: ByteArray = createState()) {
    companion object {
        init {
            System.loadLibrary("autograph")
        }

        private external fun autographStateSize(): Int

        private external fun autographHelloSize(): Int

        private external fun autographSafetyNumberSize(): Int

        private external fun autographSecretKeySize(): Int

        private external fun autographSignatureSize(): Int

        private external fun autographIndexSize(): Int

        private external fun autographSizeSize(): Int

        private external fun autographCiphertextSize(plaintextSize: Int): Int

        private external fun autographPlaintextSize(ciphertextSize: Int): Int

        private external fun autographSessionSize(state: ByteArray): Int

        private external fun autographReadIndex(index: ByteArray): Int

        private external fun autographReadSize(size: ByteArray): Int

        private external fun autographUseKeyPairs(
            publicKeys: ByteArray,
            state: ByteArray,
            identityKeyPair: ByteArray,
            ephemeralKeyPair: ByteArray,
        ): Boolean

        private external fun autographUsePublicKeys(
            state: ByteArray,
            publicKeys: ByteArray,
        )

        private external fun autographAuthenticate(
            safetyNumber: ByteArray,
            state: ByteArray,
        ): Boolean

        private external fun autographKeyExchange(
            signature: ByteArray,
            state: ByteArray,
            isInitiator: Boolean,
        ): Boolean

        private external fun autographVerifyKeyExchange(
            state: ByteArray,
            signature: ByteArray,
        ): Boolean

        private external fun autographEncryptMessage(
            ciphertext: ByteArray,
            index: ByteArray,
            state: ByteArray,
            plaintext: ByteArray,
        ): Boolean

        private external fun autographDecryptMessage(
            plaintext: ByteArray,
            size: ByteArray,
            index: ByteArray,
            state: ByteArray,
            ciphertext: ByteArray,
        ): Boolean

        private external fun autographCertifyData(
            signature: ByteArray,
            state: ByteArray,
            data: ByteArray,
        ): Boolean

        private external fun autographCertifyIdentity(
            signature: ByteArray,
            state: ByteArray,
        ): Boolean

        private external fun autographVerifyData(
            state: ByteArray,
            data: ByteArray,
            publicKey: ByteArray,
            signature: ByteArray,
        ): Boolean

        private external fun autographVerifyIdentity(
            state: ByteArray,
            publicKey: ByteArray,
            signature: ByteArray,
        ): Boolean

        private external fun autographCloseSession(
            key: ByteArray,
            ciphertext: ByteArray,
            state: ByteArray,
        ): Boolean

        private external fun autographOpenSession(
            state: ByteArray,
            key: ByteArray,
            ciphertext: ByteArray,
        ): Boolean

        fun createState(): ByteArray = ByteArray(autographStateSize())

        private fun createHello(): ByteArray = ByteArray(autographHelloSize())

        private fun createSafetyNumber(): ByteArray = ByteArray(autographSafetyNumberSize())

        private fun createSecretKey(): ByteArray = ByteArray(autographSecretKeySize())

        private fun createSignature(): ByteArray = ByteArray(autographSignatureSize())

        private fun createIndex(): ByteArray = ByteArray(autographIndexSize())

        private fun createSize(): ByteArray = ByteArray(autographSizeSize())

        private fun createCiphertext(plaintext: ByteArray): ByteArray {
            val size = autographCiphertextSize(plaintext.size)
            return ByteArray(size)
        }

        private fun createPlaintext(ciphertext: ByteArray): ByteArray {
            val size = autographPlaintextSize(ciphertext.size)
            return ByteArray(size)
        }

        private fun createSessionCiphertext(state: ByteArray): ByteArray {
            val sessionSize = autographSessionSize(state)
            val size = autographCiphertextSize(sessionSize)
            return ByteArray(size)
        }

        private fun readIndex(index: ByteArray): Int {
            return autographReadIndex(index)
        }

        private fun resizePlaintext(
            plaintext: ByteArray,
            plaintextSize: ByteArray,
        ): ByteArray {
            val size = autographReadSize(plaintextSize)
            val bytes = ByteArray(size)
            plaintext.copyInto(bytes, 0, 0, size)
            return bytes
        }
    }

    fun useKeyPairs(
        identityKeyPair: ByteArray,
        ephemeralKeyPair: ByteArray,
    ): ByteArray {
        val publicKeys = createHello()
        val success = autographUseKeyPairs(publicKeys, state, identityKeyPair, ephemeralKeyPair)
        if (!success) {
            throw RuntimeException("Initialization failed")
        }
        return publicKeys
    }

    fun usePublicKeys(publicKeys: ByteArray) {
        autographUsePublicKeys(state, publicKeys)
    }

    fun authenticate(): ByteArray {
        val safetyNumber = createSafetyNumber()
        val success = autographAuthenticate(safetyNumber, state)
        if (!success) {
            throw RuntimeException("Authentication failed")
        }
        return safetyNumber
    }

    fun keyExchange(isInitiator: Boolean): ByteArray {
        val signature = createSignature()
        val success = autographKeyExchange(signature, state, isInitiator)
        if (!success) {
            throw RuntimeException("Key exchange failed")
        }
        return signature
    }

    fun verifyKeyExchange(signature: ByteArray) {
        val success = autographVerifyKeyExchange(state, signature)
        if (!success) {
            throw RuntimeException("Key exchange verification failed")
        }
    }

    fun encrypt(plaintext: ByteArray): Pair<Int, ByteArray> {
        val ciphertext = createCiphertext(plaintext)
        val index = createIndex()
        val success = autographEncryptMessage(ciphertext, index, state, plaintext)
        if (!success) {
            throw RuntimeException("Encryption failed")
        }
        return Pair(readIndex(index), ciphertext)
    }

    fun decrypt(ciphertext: ByteArray): Pair<Int, ByteArray> {
        val plaintext = createPlaintext(ciphertext)
        val size = createSize()
        val index = createIndex()
        val success = autographDecryptMessage(plaintext, size, index, state, ciphertext)
        if (!success) {
            throw RuntimeException("Decryption failed")
        }
        return Pair(readIndex(index), resizePlaintext(plaintext, size))
    }

    fun certifyData(data: ByteArray): ByteArray {
        val signature = createSignature()
        val success = autographCertifyData(signature, state, data)
        if (!success) {
            throw RuntimeException("Certification failed")
        }
        return signature
    }

    fun certifyIdentity(): ByteArray {
        val signature = createSignature()
        val success = autographCertifyIdentity(signature, state)
        if (!success) {
            throw RuntimeException("Certification failed")
        }
        return signature
    }

    fun verifyData(
        data: ByteArray,
        publicKey: ByteArray,
        signature: ByteArray,
    ): Boolean {
        return autographVerifyData(state, data, publicKey, signature)
    }

    fun verifyIdentity(
        publicKey: ByteArray,
        signature: ByteArray,
    ): Boolean {
        return autographVerifyIdentity(state, publicKey, signature)
    }

    fun close(): Pair<ByteArray, ByteArray> {
        val key = createSecretKey()
        val ciphertext = createSessionCiphertext(state)
        val success = autographCloseSession(key, ciphertext, state)
        if (!success) {
            throw RuntimeException("Failed to close session")
        }
        return Pair(key, ciphertext)
    }

    fun open(
        key: ByteArray,
        ciphertext: ByteArray,
    ) {
        val success = autographOpenSession(state, key, ciphertext)
        if (!success) {
            throw RuntimeException("Failed to open session")
        }
    }
}
